apps/web/pages/api/auth/forgot-password.ts

```diff
@@ -42,70 +46,16 @@ async function handler(req: NextApiRequest, res: NextApiResponse) {
   if (!limit.success) {
     throw new Error(ErrorCode.RateLimitExceeded);
   }
-
+  // ensure no errors thrown by passwordResetRequest get leaked to the frontend.
   try {
-    const maybeUser = await prisma.user.findUnique({
-      where: {
-        email: email.data,
-      },
-      select: {
-        name: true,
-        identityProvider: true,
-        email: true,
-      },
-    });
-
-    if (!maybeUser) {
-      // Don't leak information about whether an email is registered or not
-      return res
-        .status(200)
-        .json({ message: "If this email exists in our system, you should receive a Reset email." });
-    }
-
-    const maybePreviousRequest = await prisma.resetPasswordRequest.findMany({
-      where: {
-        email: maybeUser.email,
-        expires: {
-          gt: new Date(),
-        },
-      },
-    });
-
-    let passwordRequest: ResetPasswordRequest;
-
-    if (maybePreviousRequest && maybePreviousRequest?.length >= 1) {
-      passwordRequest = maybePreviousRequest[0];
-    } else {
-      const expiry = dayjs().add(PASSWORD_RESET_EXPIRY_HOURS, "hours").toDate();
-      const createdResetPasswordRequest = await prisma.resetPasswordRequest.create({
-        data: {
-          email: maybeUser.email,
-          expires: expiry,
-        },
-      });
-      passwordRequest = createdResetPasswordRequest;
-    }
-
-    const resetLink = `${process.env.NEXT_PUBLIC_WEBAPP_URL}/auth/forgot-password/${passwordRequest.id}`;
-    await sendPasswordResetEmail({
-      language: t,
-      user: maybeUser,
-      resetLink,
-    });
-
-    /** So we can test the password reset flow on CI */
-    if (process.env.NEXT_PUBLIC_IS_E2E) {
-      return res.status(201).json({
-        message: "If this email exists in our system, you should receive a Reset email.",
-        resetLink,
-      });
-    } else {
-      return res
-        .status(201)
-        .json({ message: "If this email exists in our system, you should receive a Reset email." });
-    }
+    await passwordResetRequest(email.data, req.body.language ?? "en");
+    // By adding a random delay any attacker is unable to bruteforce existing email addresses.
+    const delayInMs = Math.floor(Math.random() * 2000) + 1000;
+    return await delay(() => {
+      return res.status(201).json({ message: t("password_reset_email_sent") });
+    }, delayInMs);
```

Comment:
I had a doubt. Why do we need a delay actually? Since we are already awaiting for the mail to be sent.

Reply:
this is in case of early returns; function execution time can be timed and by timing this you can figure out whether a user exists or not, adding a random delay prevents this.